### DISCUSSION TITLE
Is AtLeast working as expected?

### DISCUSSION THREAD

**Discussion by kiliancs** [prisma/prisma#24755] [Q&A]
### Question

I want to understand if we are using `AtLeast` correctly. The way it's named—and I don't see any comments clarifying—it should make _at least_ one of the keys passed in its second argument required. I read this [explanation](https://github.com/prisma/prisma/issues/20619#issuecomment-1761638159) by @SevInf:

> Hey, everyone. I believe the confusion here comes from misreading what `Prisma.AtLeast` means.
> 
> ```ts
> Prisma.AtLeast<{ id?: number, email?: string }, 'id' | 'email' >
> ```
> 
> This does not mean that both `id` and `email` are required. This means that _either_ `id` or `email` is required. So, the following inputs are valid:
> 
> - specifying only `id`
> - specifying only `email`
> - specifying both `id` and `email`
> 
> The only thing that is disallowed is omitting both fields.

So, with that in mind, let's consider a case where we have only one field to require.

```ts
export type UniqueWhereInput = AtLeast<{
    id?: string;
  }, "id">;

declare const uniqueWhereInput: UniqueWhereInput;
//                     ^? | ({ id: string; } & { id?: string; })
//                        | ({ id: string | undefined; } & { id?: string; })
//                                            ↑↑↑↑↑
//                                  Should not allow undefined

export const foo: { id: string; } = uniqueWhereInput;
//           ↑↑↑
// Should be assignable, but isn't. Errors with:
// Type 'UniqueWhereInput' is not assignable to type '{ id: string; }'.
//   Type '{ id: string | undefined; } & { id?: string | undefined; }' is not assignable to type '{ id: string; }'.
//     Types of property 'id' are incompatible.
//       Type 'string | undefined' is not assignable to type 'string'.
//         Type 'undefined' is not assignable to type 'string'.(2322)
```

In the [TS Playground](https://www.typescriptlang.org/play/?noUncheckedIndexedAccess=true&noUnusedLocals=true&noUnusedParameters=true&isolatedDeclarations=true&useUnknownInCatchVariables=true&exactOptionalPropertyTypes=false&noImplicitOverride=true&noPropertyAccessFromIndexSignature=true&noFallthroughCasesInSwitch=true#code/PTAEGMEMFcGcFNQBcCeAHetwCcCWalQA7Ae0KRNHgA81IiATZdTUeptbTebAN0SKQAtpgBQqDKAByJAKK12AHgAqAPlABeUMqrUk8RrFDQiAa1IB3IqAD820AC5i8ftgDco8S1ABBJABl4SFgkRQB5XX1DUBIAIwAreHAkABpQAGlIgwYjELwiAHN1LRl5OkZFUVBQCJoonOMzSyIq6ttQAB9QAApMuuyjU3gUEgAzGvaAb1AAbQAFUFxrdIBdJzD5ldAAX1AAMgn1gEpWts7QSfnF6yGR8Yjg0AX+6My7TKciFx4VgFobdabXYHMKnRzOVyqDyiGhoEjYcjeACqRFwAEdoPAAOoACx48AAkkQ0NBCFo-IFgqFJmDcAwAaA8ksCh5qts0gAiOkc1SeCSIFHozFEklIAlMLSCjHYvFcEWkmYAcjpipWHhAbQAejZRBrqgBlHEkaAAGyYpEIkBNJpIFkaDHgoyW8CYZ1AgGk-z1e0CeB3gE2QLgQEhEEKNIUy-HypBOKWY3FR4mk9VgQ3Gs3EMhUbDYeG6sBut2AZMISzDaPDCOAQ2HgrBcAUiCIiEhlJgkABGJzTOlOJmFNw7TTh6UJuVJpDQoA).

Would appreciate some clarity. Thank you in advance!

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by SevInf** [comment]
`Prisma.AtLeast` is an internal type that is not meant to be used for your own applications. I only clarified it usage in the issue because people were incorrectly assuming what it does. Could your reproduce the issues using public Prisma APIs? If so, can you open a bug report.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
